File: api/scala/redbus/src/main/scala/sergiusd/redbus/producer/Option.scala
package sergiusd.redbus.producer

import sergiusd.redbus.api.ProduceRequest

import java.time.ZonedDateTime

object Option {

  type Fn = ProduceRequest => ProduceRequest

  def WithKey(key: String): Fn = {
    consumer => consumer.copy(key = key)
  }

  def WithVersion(version: Long): Fn = {
    consumer => consumer.copy(version = version)
  }

  def WithIdempotencyKey(key: String): Fn = {
    consumer => consumer.copy(idempotencyKey = key)
  }

  def WithTimestamp(timestamp: ZonedDateTime): Fn = {
    consumer => consumer.copy(timestamp = timestamp.toOffsetDateTime.toString)
  }
}

File: api/golang/producer/option.go
package producer

import (
	"github.com/prokraft/redbus/api/golang/pb"
)

type OptionFn = func(c *pb.ProduceRequest)

func WithIdempotencyKey(key string) OptionFn {
	return func(r *pb.ProduceRequest) {
		r.IdempotencyKey = key
	}
}

func WithKey(key string) OptionFn {
	return func(r *pb.ProduceRequest) {
		r.Key = key
	}
}
